pkg/util/util.go

```diff
@@ -98,3 +128,96 @@ func AddAnnotation(item runtime.Unstructured, annotation, value string) {
 
 	metadata.SetAnnotations(annotations)
 }
+
+func IsVMIPaused(vmi *kvv1.VirtualMachineInstance) bool {
+	for _, c := range vmi.Status.Conditions {
+		if c.Type == kvv1.VirtualMachineInstancePaused {
+			return true
+		}
+	}
+
+	return false
+}
+
+// This is assigned to a variable so it can be replaced by a mock function in tests
+var IsDVExcludedByLabel = func(namespace, dvName string) (bool, error) {
+	client, err := GetCDIclientset()
+	if err != nil {
+		return false, err
+	}
+
+	dv, err := (*client).CdiV1beta1().DataVolumes(namespace).Get(context.TODO(), dvName, metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	labels := dv.GetLabels()
+	if labels == nil {
+		return false, nil
+	}
+
+	label, ok := labels["velero.io/exclude-from-backup"]
+	return ok && label == "true", nil
+}
+
+// This is assigned to a variable so it can be replaced by a mock function in tests
+var IsPVCExcludedByLabel = func(namespace, pvcName string) (bool, error) {
+	client, err := GetK8sClient()
+	if err != nil {
+		return false, err
+	}
+
+	pvc, err := (*client).CoreV1().PersistentVolumeClaims(namespace).Get(context.TODO(), pvcName, metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	labels := pvc.GetLabels()
+	if labels == nil {
+		return false, nil
+	}
+
+	label, ok := labels["velero.io/exclude-from-backup"]
+	return ok && label == "true", nil
+}
+
+// RestorePossible returns false in cases when restoring a VM would not be possible due to missing objects
+func RestorePossible(volumes []kvv1.Volume, backup *v1.Backup, namespace string, skipVolume func(volume kvv1.Volume) bool, log logrus.FieldLogger) (bool, error) {
+	// Restore will not be possible if a DV or PVC volume outside VM's DVTemplates is not backed up
+	for _, volume := range volumes {
+		if volume.VolumeSource.DataVolume != nil && !skipVolume(volume) {
+			if !IsResourceInBackup("datavolume", backup) {
+				log.Infof("DataVolume not included in backup")
+				return false, nil
+			}
+
+			excluded, err := IsDVExcludedByLabel(namespace, volume.VolumeSource.DataVolume.Name)
+			if err != nil {
+				return false, err
+			}
+			if excluded {
+				log.Infof("DataVolume not included in backup")
+				return false, nil
+			}
+		}
+
+		if volume.VolumeSource.PersistentVolumeClaim != nil {
+			if !IsResourceInBackup("persistentvolumeclaims", backup) {
+				log.Infof("PVC not included in backup")
+				return false, nil
+			}
+
+			excluded, err := IsPVCExcludedByLabel(namespace, volume.VolumeSource.PersistentVolumeClaim.ClaimName)
+			if err != nil {
+				return false, err
+			}
+			if excluded {
+				log.Infof("PVC not included in backup")
+				return false, nil
+			}
+		}
+		// TODO: what about other types of volumes?
```

Comment:
what other types?

Reply:
Other types of `VolumeSource`s, like Host Disks or Secret Volumes.